openlibrary/plugins/upstream/models.py

```diff
@@ -691,48 +698,118 @@ def get_sorted_editions(
         ebooks_only: bool = False,
         limit: int | None = None,
         keys: list[str] | None = None,
+        use_solr: bool = False,
     ) -> list[Edition]:
-        """
-        Get this work's editions sorted by publication year
-        :param list[str] keys: ensure keys included in fetched editions
-        """
-        db_query = {"type": "/type/edition", "works": self.key}
-        db_query['limit'] = limit or 10000  # type: ignore[assignment]
-
-        edition_keys = []
-        if ebooks_only:
-            if self._solr_data:
-                from openlibrary.book_providers import get_book_providers
-
-                # Always use solr data whether it's up to date or not
-                # to determine which providers this book has
-                # We only make additional queries when a
-                # trusted book provider identifier is present
-                for provider in get_book_providers(self._solr_data):
-                    query = {**db_query, **provider.editions_query}
-                    edition_keys += web.ctx.site.things(query)
-            else:
-                db_query["ocaid~"] = "*"
+        def get_solr_editions():
+            work = self._solr_data
+            edition_keys = work.get('edition_key') or []
+            edition_queries = [f'key:"/books/{key}"' for key in edition_keys]
+
+            ed_results = run_solr_query(
+                EditionSearchScheme(),
+                {'q': ' OR '.join(edition_queries)},
+                rows=limit,
+                fields=[
+                    'ia',
+                    'key',
+                    'title',
+                    'subtitle',
+                    'language',
+                    'isbn',
+                    'publish_date',
+                    'publisher',
+                ],
+                user_query=False,
+            )
 
-        if not edition_keys:
-            solr_is_up_to_date = (
-                self._solr_data
-                and self._solr_data.get('edition_key')
-                and len(self._solr_data.get('edition_key')) == self.edition_count
+            _editions = []
+            if ed_results.docs:
+                for solr_edition in ed_results.docs:
+                    ed = web.storify(solr_edition)
```

Comment:
Yeah this coercion is a little annoying, alas! Let's brainstorm about this perhaps? I wonder if we can use a [`Protocol`](https://peps.python.org/pep-0544/), and have our infogami `Edition` and create a new class in this file, `class SolrEditionWrapper` conform to that protocol? That should help ensure they have the same methods necessary for some of the basic things (isbns, etc), while still letting them have the unique fields that one might sometimes have?

Although we are also planning to update solr to have all the missing fields... so maybe it's easier to update the edition table html to always assume a solr doc?

Reply:
Or maybe the simplest approach is to do something like in SearchResultWork:

https://github.com/internetarchive/openlibrary/blob/29208c70acafe439f1ec0a3f153cc747beed56da/openlibrary/macros/SearchResultsWork.html#L5-L10